gpflow/models/training_interface.py

```diff
@@ -0,0 +1,62 @@
+import abc
+import collections
+from typing import Callable, Optional, Tuple, TypeVar
+
+
+import tensorflow as tf
+import numpy as np
+
+InputData = tf.Tensor
+OutputData = tf.Tensor
+RegressionData = Tuple[InputData, OutputData]
+Data = TypeVar("Data", RegressionData, InputData)
+
+
+class TrainingInterface:
+    @abc.abstractmethod
+    def training_loss(self, *args, **kwargs) -> tf.Tensor:
+        """
+        Specify a training loss for this model
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def training_loss_closure(self, *args, **kwargs):
+        """
+        Return a closure with a training loss for this module
+        """
+        raise NotImplementedError
+
+
+class InternalDataTrainingInterface(TrainingInterface):
+    def training_loss(self):
+        """
+        Specify a training loss for this model
+        """
+        raise NotImplementedError
+
+    def training_loss_closure(self, jit=True) -> Callable[[], tf.Tensor]:
+        if jit:
+            return tf.function(self.training_loss)
+        return self.training_loss
+
+
+class ExternalDataTrainingInterface(TrainingInterface):
+    def training_loss(self, data):
+        """
+        Specify a training loss for this model
+        """
+        raise NotImplementedError
+
+    def training_loss_closure(self, data: Data, jit=True) -> Callable[[], tf.Tensor]:
```

Comment:
type signature doesn't agree with the use in the closure() below...?
```suggestion
    def training_loss_closure(self, data: Union[Data, Iterator], jit=True) -> Callable[[], tf.Tensor]:
```